toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -68,6 +68,16 @@ func main() {
 		logger.Log.Fatalf("unable to read toolchain manifest file '%s': %s", *toolchainManifest, err)
 	}
 
+	dependencyGraph.InitLookup()
+
+	/*If there is an existing runNode and then there is one more remote node, donot throw dup error, instead replace the run node with remote node*/
+	for _, n := range graph.NodesOf(dependencyGraph.Nodes()) {
+		pkgNode := n.(*pkggraph.PkgNode)
+		if pkgNode.Type == pkggraph.TypeRemote {
+			dependencyGraph.AddRemoteToLookup(pkgNode, true)
```

Comment:
>dependencyGraph.AddRemoteToLookup(pkgNode, true) [](http://example.com/codeflow?start=3&length=48)

Another comment related to the lookup table being part of the internal implementation: we shouldn't be exposing such functions as public and make it caller's responsibility to take care of the lookup table's state. The graph is responsible for keeping it accurate and up-to-date. Could you say when are we modifying the graph in such way, that this code was now required here? I think we should take care of cleaning up the lookup table at the time of that modification, doing it as part of the graph's internal code and keep the caller out of it.

Reply:
I could have added this as part of initLookup() with an extra parameter to enable adding remotes to lookup. But I chose not to edit the existing initLookup function and put this functionality separate as this is used only in graphpkgfetcher.

Reply:
When there is a cycle and it can be resolved by upstream RPMs, then I'm adding remote Node to the graph, replacing the runNode in the cycle with remoteNode. So, in this case, both run and remote nodes exist for the same node. Due to fresh initialization of lookupTable from the graph by the graphpkgfetcher, sometimes runNode is populated and and sometimes remoteNode is populated in the lookup table. To address this inconsistency, I am replacing the runNode with remoteNode in the lookupTable, so that it gets fetched by the graphpkgfetcher without fail.  This replacing is what the "addRemotetoLookup" does. If I have to not expose this functionality, I need to make the lookupTable function public and send a parameter to do this during initilization of lookupTable.

Reply:
I have a question @dmcilvaney and @PawelWMS Why does graphpkgfetcher have to refer the LookupTable to fetch remote Nodes? It can directly refer the graph using graph.AllNodes, check if its TypeRemote and fetch these nodes. If we do this, This modification in lookup would not be needed. I think this is the best way.

Reply:
Walking the entire graph via `AllNodes()` can be fairly time consuming since we need to check and covert ever node into a version range struct and do a comparison. The lookup allows us to quickly find a node that satisfies a given node+version requirement. The lookup is also sorted in such a way that it will preferentially give the highest version back for a given package if there are multiple copies.

Reply:
This feels like the better option here is to improve `validateNodeForLookup()` and `addToLookup()` to support adding a distinct remote node without trying to re-use the existing local run node.

Reply:
Replacing the runNode with remoteNode in lookupTable is done only in the graphpkgFetcher, and lookupTable is built from the graph at every build stage, So the lookup Table at this stage will not affect any other build stages. 

Also, in graphpkgFetcher, lookupTable is used only to find and resolve remote Nodes  and there is no dependency on the run Node that is removed from the lookupTable. 
So, there should be no issue with this method.

Let me know if you think this approach will affect anything in the graphpkgFetcher?

Reply:
I'm still very learly of exposing the lookup table details to an external caller. This still feels like something that should be hanlded better inside the graph package itself rather than requiring an external call to force it.

Reply:
Could we also add a significant comment here? It is very unclear when reading this WHY we want to replace these nodes when reading the code. Also, do we want to replace the node, or add a second copy? Won't we just generate an error next time we try to load the graph file? I would expect `validateNodeForLookup()` to complain if we have both a remote and local run node stored in the same graph.

Reply:
Ah, I found the bug that was allowing this state to occur.
In `initLookup()` we have:
```go
// initLookup initializes the run and build node lookup table
func (g *PkgGraph) initLookup() {
	g.nodeLookup = make(map[string][]*LookupNode)

	// Scan all nodes, start with only the run nodes to properly initialize the lookup structures
	// (they always expect a run node to be present)
	for _, n := range graph.NodesOf(g.Nodes()) {
		pkgNode := n.(*PkgNode)
		if pkgNode.Type == TypeLocalRun || pkgNode.Type == TypeRemoteRun {
			g.addToLookup(pkgNode, true)
		}
	}

	// Now run again for any build nodes, or other nodes we want to track
	for _, n := range graph.NodesOf(g.Nodes()) {
		pkgNode := n.(*PkgNode)
		if pkgNode.Type != TypeLocalRun && pkgNode.Type != TypeRemoteRun {
			g.addToLookup(pkgNode, true)
		}
	}
```

Note we don't check the errors from `g.addToLookup(pkgNode, true)`. I think the idea here was we would have validated all input while adding nodes to the graph.  We should consider adding error checking here.

Reply:
My position is: These changes should continue to work even if we add the error checking to this code. I think the forceful manipulation of the graph lookup cache will no longer be viable here because we will now refuse to load the invalid graph from file.